mlir/lib/Dialect/Rock/Transforms/RemoveRedundantCasts.cpp

```diff
@@ -0,0 +1,510 @@
+// performance by reducing superfluous operations and memory traffic.
+//
+//===----------------------------------------------------------------------===//
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Rock/IR/AmdArchDb.h"
+#include "mlir/Dialect/Rock/IR/GetRockInfo.h"
+#include "mlir/Dialect/Rock/Passes.h"
+#include "mlir/Dialect/Rock/utility/builderUtils.h"
+#include "mlir/Dialect/Rock/utility/loweringUtils.h"
+#include "mlir/Dialect/Rock/utility/transformMapUtils.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace rock {
+#define GEN_PASS_DEF_ROCKREMOVEREDUNDANTCASTSPASS
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-remove-redundant-casts"
+
+using namespace mlir;
+using namespace mlir::arith;
+using namespace mlir::rock;
+
+namespace {
+
+// Pattern to remove redundant dtype -> newDtype -> dtype cast chains
+struct RemoveRedundantTruncExtfPattern
+    : public OpRewritePattern<linalg::GenericOp> {
+  func::FuncOp func;
+
+  RemoveRedundantTruncExtfPattern(MLIRContext *context, func::FuncOp func)
+      : OpRewritePattern<linalg::GenericOp>(context), func(func) {}
+
+  LogicalResult matchAndRewrite(linalg::GenericOp generic,
+                                PatternRewriter &rewriter) const override {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Looking at generic: " << generic.getLoc() << "\n");
+
+    // Check if the generic operation only contains an ext operation.
+    if (!(isGenericWithSingleOp<arith::ExtFOp>(generic) ||
+          isGenericWithSingleOp<arith::ExtSIOp>(generic) ||
+          isGenericWithSingleOp<arith::ExtUIOp>(generic))) {
+      LLVM_DEBUG(llvm::dbgs() << "Generic doesn't contain only ext\n");
+      return failure();
+    }
+
+    bool changed = false;
+
+    auto extInput = getExtInput(generic);
+    if (!extInput)
+      return failure();
+
+    // We have checks in getExtInput to ensure that the input is not a
+    // BlockArgument, so we can assume that it has a defining op.
+    auto input = extInput.getDefiningOp();
+
+    if (isa<RockGemmWrapperInterface>(input) ||
+        isa<RockGemmGemmWrapperInterface>(input)) {
+      // Now we need to check the input of the cast (linalg.generic with ext)
+      // to see if it is a RockOp that makes use of MFMA. MFMA in rocmlir does
+      // accumulation in higher precision, so if our rock op is using MFMA and
+      // the return type isn't F32 or I32 (highest level of precision), then it
+      // means that there will be a trunc op inserted in RockGemmToGridwise
+      // that will potentially be redundant.
+      changed |= handleRockGemmWrapper(input, generic, rewriter);
```

Comment:
```suggestion
     return handleRockGemmWrapper(input, generic, rewriter);
```

make `handleRockGemmWrapper` return LogicalResult() directly instead of relying on `changed` which seems unnecessary given each branch of IfThenElse is returning.